datastore-design/go-grpc-demo/protobuf/datastore.proto

```diff
@@ -49,6 +54,27 @@ message AddTSResponse {
 
 //---------------------------------------------------------------------------
 
+message Polygon { // horizontal area
+  // three or more points with an implicit edge between the first and last point
+  repeated Point points = 1;
+}
+
+message FindTSRequest {
+  repeated string station_ids = 1; // if specified, only return time series matching any of these station IDs
+  repeated string param_ids = 2; // if specified, only return time series matching any of these param IDs
+  Polygon inside = 3; // if specified, only return time series located inside this area
+  // TODO: add search filters for other metadata
+}
+
+message FindTSResponse {
+  int32 status = 1;
+  string error = 2; // empty on success
+  repeated int64 ids = 3; // matching time series IDs
+  repeated TSMetadata metadata = 4; // corresponding metadata
```

Comment:
I am not a fan of having two arrays where the indices have to match up to find the (id, metadata) pair. Can we have some kind of "tuple" structure which holds the id and the TSMetadata, and then have `repeated` on that?

Reply:
I agree! Eventually a "parallel arrays" idiom could be reintroduced for performance if need be. But let's try to avoid premature optimization. Fix coming up ...